R/estimateDiversity.R

```diff
@@ -239,17 +344,66 @@ setMethod("estimateDiversity", signature = c(x = "SummarizedExperiment"),
     vegan::fisher.alpha(t(mat))
 }
 
+.calc_faith <- function(mat, tree, ...){
+
+    # Gets name of the samples
+    samples <- colnames(mat)
+    #taxa <- rownames(mat)
+
+    # Repeats taxa as many times there are samples, i.e. get all the taxa that are
+    # analyzed in each sample
+    taxa <- rep(rownames(mat), length(samples))
+
+    # Gets those taxa that are present/absent in each sample. Gets one big list that combines
+    # taxa from all the samples.
+    present_combined <- taxa[ mat[, samples] > 0 ]
+    absent_combined <- taxa[ mat[, samples] == 0 ]
+
+    # colSums(mat > 0) # Gets how many taxa there are in each sample
+    # as.vector(cumsum(colSums(mat > 0)+1)) # Determines indices of samples' first taxa
+    # seq_along(present_combined) %in% as.vector(cumsum(colSums(mat > 0)+1)) # Gets true if splitting point
+    # as.factor(cumsum((seq_along(present_combined)-1) %in% as.vector(cumsum(colSums(mat > 0)))))) # Determines which taxa belongs to which sample
+    # Split() assigns taxa to right samples based on their number that they got from previous step
+    # unname() deletes unnecessary names
+    present <- unname(split(present_combined, as.factor(cumsum((seq_along(present_combined)-1) %in% as.vector(cumsum(colSums(mat > 0)))))))
+    absent <- unname(split(absent_combined, as.factor(cumsum((seq_along(absent_combined)-1) %in% as.vector(cumsum(colSums(mat == 0)))))))
```

Comment:
For clarity, it is always better split the factor generation from actually using it. Also the line length does not exceed 80/100 characters:

```r
split_present <- as.vector(cumsum(colSums(mat > 0)))
split_absent <- as.vector(cumsum(colSums(mat == 0)))
split_present <- as.factor(cumsum((seq_along(present_combined)-1) %in% split_present))
split_absent <- as.factor(cumsum((seq_along(absent_combined)-1) %in% split_absent))
present <- unname(split(present_combined, split_present))
absent <- unname(split(absent_combined, split_absent))
```

Reply:
Good point. Much easier to follow the code, if user wants to

Changed

Reply:
And for yourself in two years 😄

Reply:
Well that's also true :D